KAFKA-18206: EmbeddedKafkaCluster must set features

brandboat opened the pull request:
related to KAFKA-18206, set features in EmbeddedKafkaCluster in both streams and connect module, note that this PR also fix potential transaction with empty records in sendPrivileged method as transaction version 2 doesn't allow this kind of scenario.


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

brandboat committed:
KAFKA-18206: EmbeddedKafkaCluster must set features

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18206

mjsax commented:
> note that this PR also fix potential transaction with empty records in sendPrivileged method as transaction version 2 doesn't allow this kind of scenario.

Should we split this into it's own PR?

brandboat commented:
> Should we split this into it's own PR?

Hi @mjsax, sorry I forgot to mention that this was handling in a different issue, here is the PR: https://github.com/apache/kafka/pull/18448

jolshan commented:
We have merged 18448 so we should be able to proceed here :)

brandboat committed:
Update

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18206

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18206

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18206

brandboat commented:
I'm not sure why, but it seems like a lot of tests have become more flaky than before in this PR. One failed test in CI (https://github.com/apache/kafka/actions/runs/12810813472) caught my attention: MirrorConnectorsIntegrationTransactionsTest #testOffsetTranslationBehindReplicationFlow. The error message is:
```
java.lang.RuntimeException: java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.InvalidTxnStateException: The producer attempted a transactional operation in an invalid state.
```
I expected this failed test should to be mitigated after merging [PR #18448](https://github.com/apache/kafka/pull/18448), so I'll spend some time investigating this.

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18206

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18206

brandboat commented:
This https://github.com/apache/kafka/pull/18189#issuecomment-2596227623 seems due to TV_2 return incorrect error, and should be handled correctly in https://github.com/apache/kafka/pull/18604.

jolshan commented:
@brandboat I just merged my PR and will restart the tests here. You may need to rebase, but we can try as is for now.

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18206

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18206

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18206

brandboat commented:
Bunch of ci failed in this PR since ClearElrRecord isn't corrected handled, but will be fixed after https://github.com/apache/kafka/pull/18708 merged.

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18206

jolshan commented:
The build is looking better! Maybe we just need to look in to the `KRaftClusterTest` failure?

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18206

brandboat commented:
> The build is looking better! Maybe we just need to look in to the KRaftClusterTest failure?

Yes, I'm currently investigating the root cause of this test failure. Thanks for the heads-up!

brandboat committed:
Fix testAdminClientApisAuthenticationFailure

after ec49a60e4f690f311060045ee1feb20d99837f97, if ELR is enabled,
a cluster-level min.insync.replicas will be set to 1, and this PR
aims to update all feature version to the latest.

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18206

brandboat commented:
Finally got an AC 🎉 , perhaps we can move on?

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18206

jolshan commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -696,7 +696,8 @@ class KRaftClusterTest {
           ("log.roll.ms", "1234567"),
-          ("max.connections.per.ip", "60"))), exhaustive = true)
+          ("max.connections.per.ip", "60"),
+          ("min.insync.replicas", "1"))), exhaustive = true)
```
Why did we make this change?

brandboat replied:
after https://github.com/apache/kafka/commit/ec49a60e4f690f311060045ee1feb20d99837f97, if ELR is enabled, a cluster-level min.insync.replicas will be set to 1, and this PR will enable ELR by default in all integration tests that use TestKitNodes.java

jolshan commented:
I'm also curious why the tests are returning 1 if there are just flaky tests. Has this always been the case?

brandboat commented:
> I'm also curious why the tests are returning 1 if there are just flaky tests. Has this always been the case?

Hi @jolshan, pardon me, did you mean EagerConsumerCoordinatorTest.testOutdatedCoordinatorAssignment? If so, yes it it a flaky test, and we already have a JIRA ticket for it: https://issues.apache.org/jira/browse/KAFKA-15900

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18206

jolshan reviewed (approved):
Thanks for your persistence here @brandboat !

jolshan commented:
Also pushed to 4.0. Thanks!
